src/org/traccar/api/resource/DeviceResource.java

```diff
@@ -42,20 +44,31 @@
 
     @GET
     public Collection<Device> get(
-            @QueryParam("all") boolean all, @QueryParam("userId") long userId) throws SQLException {
-        if (all) {
-            if (Context.getPermissionsManager().isAdmin(getUserId())) {
-                return Context.getDeviceManager().getAllDevices();
+            @QueryParam("all") boolean all, @QueryParam("userId") long userId,
+            @QueryParam("id") List<Long> deviceIds) throws SQLException {
+        if (deviceIds.isEmpty()) {
+            if (all) {
+                if (Context.getPermissionsManager().isAdmin(getUserId())) {
+                    return Context.getDeviceManager().getAllDevices();
+                } else {
+                    Context.getPermissionsManager().checkManager(getUserId());
+                    return Context.getDeviceManager().getManagedDevices(getUserId());
+                }
             } else {
-                Context.getPermissionsManager().checkManager(getUserId());
-                return Context.getDeviceManager().getManagedDevices(getUserId());
+                if (userId == 0) {
+                    userId = getUserId();
+                }
+                Context.getPermissionsManager().checkUser(getUserId(), userId);
+                return Context.getDeviceManager().getDevices(userId);
             }
         } else {
-            if (userId == 0) {
-                userId = getUserId();
+            ArrayList<Device> devices = new ArrayList<>();
+            for (Long deviceId : deviceIds) {
+                Device device = Context.getDeviceManager().getDeviceById(deviceId);
+                Context.getPermissionsManager().checkDevice(getUserId(), device.getId());
```

Comment:
We can do permission check first before getting device. Also, there is no need for `device` variable.